src/devtools/pachdev/actions.go

```diff
@@ -0,0 +1,571 @@
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+// ctx and cancel are captured by result.Ctx(), and will not be initialized
+// if that is never called
+func (op *deployOp) Ctx() context.Context {
+	if op.ctx != nil {
+		return op.ctx
+	}
+	op.ctx, op.cancel = context.WithCancel(context.Background())
+	return op.ctx
+}
+
+func (op *deployOp) Cancel() {
+	if op.cancel != nil {
+		op.cancel()
+	}
+}
+
+// KubeClient is an accessor for op.kubeClient that lazily initializes the
+// field if it's not already set.  All access to op.kubeClient should be through
+// this method, even internally, in case it hasn't been initialized yet.
+func (op *deployOp) KubeClient() *kubernetes.Clientset {
+	if op.kubeClient != nil {
+		return op.kubeClient
+	}
+
+	// Load the Kubernetes configuration from the default location
+	config, err := clientcmd.BuildConfigFromFlags("", clientcmd.RecommendedHomeFile)
+	if err != nil {
+		op.err = errors.Wrap(err, "could not load kubernetes config")
+	}
+
+	op.kubeClient, err = kubernetes.NewForConfig(config)
+	if err != nil {
+		op.err = errors.Wrap(err, "could not create kubernetes client")
+	}
+	return op.kubeClient
+}
+
+// PachClient is an accessor for op.pachClient that lazily initializes the
+// field if it's not already set.  All access to op.kubeClient should be through
+// this method, even internally, in case it hasn't been initialized yet.
+func (op *deployOp) PachClient() *client.APIClient {
+	if op.pachClient != nil {
+		return op.pachClient
+	}
+
+	pachctlCfg := &pachctl.Config{}
+	var err error
+	op.pachClient, err = pachctlCfg.NewOnUserMachine(op.Ctx(), false)
+	if err != nil {
+		op.err = errors.Wrap(err, "could not create Pachyderm client")
+	}
+	return op.pachClient
+}
+
+// KindProvider is an accessor for op.kindProvider that lazily initializes the
+// field if it's not already set.  All access to op.kindProvider should be through
+// this method, even internally, in case it hasn't been initialized yet.
+func (op *deployOp) KindProvider() *cluster.Provider {
+	if op.kindProvider != nil {
+		return op.kindProvider
+	}
+	var kindLogger log.Logger = log.NoopLogger{}
+	if verbose {
+		// TODO(msteffen) literally no idea what good UX is here. Ideally we
+		// would log to stdout instead of stderr, but I can't access
+		// kind/pgk/internal/cli to change the logging destination. Am I holding
+		// this wrong?
+		kindLogger = cmd.NewLogger()
+	}
+	op.kindProvider = cluster.NewProvider(
+		cluster.ProviderWithLogger(kindLogger),
+		cluster.ProviderWithDocker(), // TODO(msteffen): is this necessary or appropriate?
+	)
+	return op.kindProvider
+}
+
+// helmChartSource is a helper for K8sManifest that returns the location of the
+// Pachyderm helm chart it should use.
+func (op *deployOp) helmChartSource() string {
+	if op.pachVersion == "latest" {
+		// // TODO(msteffen): this codepath has no automated tests. Maybe set up
+		// // mock GitHub? (has worked in the past, and I bet the releases endpoint
```

Comment:
That could be good.  I handle this manually now in 2 ways:

1. Just deploy an old tag with the current working copy helm chart.
2. "git checkout v2.8.2" and deploy with that helm chart.

If this thing ends up "go build"-able, then you can build the latest version of pachdev and run it against an "old" working copy.

If we want to be really consistent about using old versions, like for upgrade tests, I think pachyderm will just depend on the deployment artifacts from prior releases (i.e. a http_archive() WORKSPACE rule that downloads the helm chart for 2.7.2 or whatever we want).